tariff/pun.go

```diff
@@ -123,48 +127,67 @@ func (t *Pun) Type() api.TariffType {
 }
 
 func (t *Pun) getData(day time.Time) (api.Rates, error) {
-	// Cookie Jar zur Speicherung von Cookies zwischen den Requests
+	// Cookie Jar to store cookies between requests
 	client := request.NewClient(t.log)
 	client.Jar, _ = cookiejar.New(nil)
 
-	// Erster Request
-	uri := "https://storico.mercatoelettrico.org/It/WebServerDataStore/MGP_Prezzi/" + day.Format("20060102") + "MGPPrezzi.xml"
-	resp, err := client.Get(uri)
+	// Request the ZIP file
+	uri := "https://gme.mercatoelettrico.org/DesktopModules/GmeDownload/API/ExcelDownload/downloadzipfile?DataInizio=" + day.Format("20060102") + "&DataFine=" + day.Format("20060102") + "&Date=" + day.Format("20060102") + "&Mercato=MGP&Settore=Prezzi&FiltroDate=InizioFine"
+	req, err := http.NewRequest("GET", uri, nil)
 	if err != nil {
 		return nil, err
 	}
+
+	req.Header = http.Header{
+		"moduleid":           []string{"12103"},
+		"Referer":            []string{"https://gme.mercatoelettrico.org/en-us/Home/Results/Electricity/MGP/Download?valore=Prezzi"},
+		"sec-ch-ua-mobile":   []string{"?0"},
+		"sec-ch-ua-platform": []string{"Windows"},
+		"sec-fetch-dest":     []string{"empty"},
+		"sec-fetch-mode":     []string{"cors"},
+		"sec-fetch-site":     []string{"same-origin"},
+		"sec-gpc":            []string{"1"},
+		"tabid":              []string{"1749"},
+		"userid":             []string{"-1"},
+	}
+
+	resp, err := client.Do(req)
+	if err != nil || resp.StatusCode != http.StatusOK {
+		t.log.TRACE.Println(`PUN Zip file request failed:`, err)
+		return nil, err
+	}
 	defer resp.Body.Close()
 
-	formData, err := parseFormFields(resp.Body)
+	// Process the ZIP file
+	body, err := io.ReadAll(resp.Body)
 	if err != nil {
-		return nil, fmt.Errorf("form fields: %w", err)
+		return nil, err
 	}
 
-	redirectURL := resp.Request.URL.String()
+	zipReader, err := zip.NewReader(bytes.NewReader(body), int64(len(body)))
+	if err != nil {
+		return nil, err
+	}
 
-	// Hinzufügen der spezifizierten Parameter
-	formData.Set("ctl00$ContentPlaceHolder1$CBAccetto1", "on")
-	formData.Set("ctl00$ContentPlaceHolder1$CBAccetto2", "on")
-	formData.Set("ctl00$ContentPlaceHolder1$Button1", "Accetto")
+	if len(zipReader.File) != 1 {
+		return nil, fmt.Errorf("unexpected number of files in the ZIP archive")
+	}
 
-	// Formular senden
-	resp, err = client.PostForm(redirectURL, formData)
+	zipFile := zipReader.File[0]
+	f, err := zipFile.Open()
 	if err != nil {
-		fmt.Println("Error submitting form:", err)
 		return nil, err
 	}
-	defer resp.Body.Close()
+	defer f.Close()
 
-	// Erneuter Request auf die ursprüngliche URL
-	resp, err = client.Get(uri)
+	xmlFile, err := io.ReadAll(f)
```

Comment:
Simpler

    xls.NewDecoder(resp.Body)

Reply:
xls? it's a zip file with a xml file inside. this way just works.

Reply:
Obviously

    xml.NewDecoder(resp.Body)

Reply:
Obviously, but again, resp.Body is neither a xml nor a xls but a zip file. Instead, xml.NewDecoder should be used on f , that is the xml file inside the zip file. I used an intermediate passage to make it clearer, but I will try and modify if needed.

Reply:
Body is an io.Reader. Xml Decoder works on io.Reader. I would appreciate if you just gave it a try. Thank you.